src/ZkFold/Symbolic/Data/Ed25519.hs

```diff
@@ -0,0 +1,225 @@
+import           ZkFold.Base.Algebra.EllipticCurve.Ed25519
+import           ZkFold.Prelude                            (length)
+import           ZkFold.Symbolic.Compiler                  hiding (forceZero)
+import           ZkFold.Symbolic.Data.Bool
+import           ZkFold.Symbolic.Data.Combinators
+import           ZkFold.Symbolic.Data.Conditional
+import           ZkFold.Symbolic.Data.DiscreteField
+import           ZkFold.Symbolic.Data.Eq
+import           ZkFold.Symbolic.Data.UInt
+
+zpToEd :: (Finite (Zp p)) => Point (Ed25519 (Zp p)) -> Point (Ed25519 Void)
+zpToEd Inf         = Inf
+zpToEd (Point x y) = Point (toZp . toConstant $ x) (toZp . toConstant $ y)
+
+edToZp :: (Finite (Zp p)) => Point (Ed25519 Void) -> Point (Ed25519 (Zp p))
+edToZp Inf         = Inf
+edToZp (Point x y) = Point (fromConstant . fromZp $ x) (fromConstant . fromZp $ y)
+
+-- | Ed25519 with @UInt 256 (Zp p)@ as computational backend
+--
+instance (Finite (Zp p)) => EllipticCurve (Ed25519 (Zp p)) where
+    type BaseField (Ed25519 (Zp p)) = UInt 256 (Zp p)
+    type ScalarField (Ed25519 (Zp p)) = UInt 256 (Zp p)
+
+    inf = Inf
+
+    gen = Point
+            (fromConstant (15112221349535400772501151409588531511454012693041857206046113283949847762202 :: Natural))
+            (fromConstant (46316835694926478169428394003475163141307993866256225615783033603165251855960 :: Natural))
+
+    -- | Addition casts point coordinates to @Zp Ed25519_Base@ and adds them as if they were points on @Ed25519 Void@
+    --
+    add p1 p2 = edToZp $ add (zpToEd p1) (zpToEd p2)
+
+    -- | Multiplication casts point coordinates to @Zp Ed25519_Base@ and scale factor to @Zp Ed25519_Scalar@,
+    --   then scales the point as if it were a point on @Ed25519 Void@
+    --
+    mul s p = edToZp @p $ mul (toZp . toConstant $ s) (zpToEd @p p)
+
+instance SymbolicData a (UInt 256 (ArithmeticCircuit a)) => SymbolicData a (Point (Ed25519 (ArithmeticCircuit a))) where
+
+    -- (0, 0) is never on a Twisted Edwards curve for any curve parameters.
+    -- We can encode the point at infinity as (0, 0), therefore.
+    pieces Inf         = pieces (zero :: UInt 256 (ArithmeticCircuit a)) <> pieces (zero :: UInt 256 (ArithmeticCircuit a))
+    pieces (Point x y) = pieces x <> pieces y
+
+    restore lst
+      | length lst /= (typeSize @a @(Point (Ed25519 (ArithmeticCircuit a)))) = error "SymbolicData a (Point (Ed25519 (ArithmeticCircuit a))): wrong number of registers"
+      | otherwise =
+            bool @(Bool (ArithmeticCircuit a)) @(Point (Ed25519 (ArithmeticCircuit a))) (Point x y) Inf ((x == zero) && (y == zero))
+        where
+            (piecesX, piecesY) = splitAt (P.fromIntegral (length lst) `P.div` 2) lst
+            (x, y) = (restore piecesX, restore piecesY)
+
+    typeSize = 2 * typeSize @a @(UInt 256 (ArithmeticCircuit a))
+
+instance Eq (Bool (ArithmeticCircuit a)) (UInt 256 (ArithmeticCircuit a)) => Eq (Bool (ArithmeticCircuit a)) (Point (Ed25519 (ArithmeticCircuit a))) where
```

Comment:
With Eq it can be done, but for point arithmetic I don't see a nice way to do it because of Shrink and Extend. Maybe with foreign field operations it will be easier, as they will allow us to get rid of Shrink and Extend here.